src/clock.zig

```diff
@@ -0,0 +1,235 @@
+            const majority = interval.sources_true > @divTrunc(self.sources.len, 2);
+            if (!majority) break;
+            best = interval;
+        }
+
+        return best;
+    }
+
+    fn window_tuples(self: *Clock, clock_offset_max: u64) []Marzullo.Tuple {
+        assert(self.sources[self.replica].?.clock_offset == 0);
+        assert(self.sources[self.replica].?.one_way_delay == 0);
+        var count: usize = 0;
+        for (self.sources) |sampled, source| {
+            if (sampled) |sample| {
+                const range = @intCast(i64, sample.one_way_delay) + @intCast(i64, clock_offset_max);
+                self.marzullo_tuples[count] = Marzullo.Tuple{
+                    .source = @intCast(u8, source),
+                    .offset = sample.clock_offset - range,
+                    .bound = .lower,
+                };
+                count += 1;
+                self.marzullo_tuples[count] = Marzullo.Tuple{
+                    .source = @intCast(u8, source),
+                    .offset = sample.clock_offset + range,
+                    .bound = .upper,
+                };
+                count += 1;
+            }
+        }
+        return self.marzullo_tuples[0..count];
+    }
+};
+
+// TODO Each Clock instance should accept a Time instance that can be ticked deterministically for
+// testing purposes, and which will expose the following two timestamps:
+// This is the same pattern we use for Journal (all the logic) and Storage (the raw read/write).
+
+/// A timestamp to measure elapsed time, meaningful only on the same system, not across reboots.
+/// Always use a monotonic timestamp if the goal is to measure elapsed time.
+/// This clock is not affected by discontinuous jumps in the system time, for example if the
+/// system administrator manually changes the clock.
+fn monotonic_timestamp() u64 {
+    // The true monotonic clock on Linux is not in fact CLOCK_MONOTONIC:
+    // CLOCK_MONOTONIC excludes elapsed time while the system is suspended (e.g. VM migration).
+    // CLOCK_BOOTTIME is the same as CLOCK_MONOTONIC but includes elapsed time during a suspend.
+    // For more detail and why CLOCK_MONOTONIC_RAW is even worse than CLOCK_MONOTONIC,
+    // see https://github.com/ziglang/zig/pull/933#discussion_r656021295.
+    var ts: std.os.timespec = undefined;
+    std.os.clock_gettime(std.os.CLOCK_BOOTTIME, &ts) catch unreachable;
+    return @intCast(u64, ts.tv_sec) * @as(u64, std.time.ns_per_s) + @intCast(u64, ts.tv_nsec);
+
+    // TODO Add a monotonic guard here to verify that clock is monotonic.
+}
+
+/// A timestamp to measure real (i.e. wall clock) time, meaningful across systems, and reboots.
+/// This clock is affected by discontinuous jumps in the system time.
+fn realtime_timestamp() i64 {
+    var ts: std.os.timespec = undefined;
+    std.os.clock_gettime(std.os.CLOCK_REALTIME, &ts) catch unreachable;
+    return @intCast(i64, ts.tv_sec) * @as(i64, std.time.ns_per_s) + @intCast(i64, ts.tv_nsec);
```

Comment:
Are the `@intCast()`s here really needed? IIRC the fields of the timespec struct on linux are always signed and <= 64 bits.

Reply:
Ah yes of course. I followed `std/time.zig` for the casts for `CLOCK_BOOTTIME` and then borrowed the same style for `CLOCK_REALTIME`.

Are you happy with how we cast our wall clock nano timestamps to `i64` instead of `i128` as the std lib does? It seems we have a few hundred years regardless.

Reply:
On second thought, I guess we might need the `@intCast()` if this was on a 32-bit architecture, to avoid overflow when doing the `tv_sec * ns_per_s` multiply?

Reply:
> Are you happy with how we cast our wall clock nano timestamps to i64 instead of i128 as the std lib does? It seems we have a few hundred years regardless.

Hmm, using a `u64` as the return value would make me slightly happier, but I'm not sure if the kernel can realistically report a negative time or not. `i64` is probably fine unless the system clock is set to a few hundred years in the future.

> On second thought, I guess we might need the @intCast() if this was on a 32-bit architecture, to avoid overflow when doing the tv_sec * ns_per_s multiply?

good point about 32-bit architectures, `@as()` would be sufficent though:
```zig
return @as(i64, ts.tv_sec) * std.time.ns_per_s + ts.tv_nsec;
```

Reply:
> Hmm, using a u64 as the return value would make me slightly happier, but I'm not sure if the kernel can realistically report a negative time or not.

Yes, I see the std lib was happy with u64 for Timer, because the high bit buys a few hundred years. We still have 292 years on either side of the Unix epoch if my arithmetic is correct: `Math.pow(2,63)/1e9/60/60/24/365`

Negative time is possible. This would represent anything before the Unix epoch, e.g. pre-1970s. For example, if a sys admin sets the clock back to the 60s. 😄